pkg/controller/middleware/auth.go

```diff
@@ -122,6 +122,18 @@ func RequireAuth(cacher cache.Cacher, authProvider auth.Provider, db *database.D
 				return
 			}
 
+			// Insert realm id to context so metrics can pick it up.
+			realmID := controller.RealmIDFromSession(session)
+			// This is inserted by realm selector, so it could be 0 if no realm
+			// has been selected.
+			if realmID != 0 {
+				realm, err := db.FindRealm(realmID)
```

Comment:
this should attempt to read the realm from the cache before going to DB

Reply:
You're right, but I think this is the wrong place to do this entirely. We should set this in membership.go instead, where the realm already exists with no lookup required.